File: src/strategy.py
"""
Strategy Pattern
================

I prefer to use first class functions to implement Strategies because it makes
the code terser in general.
"""


def no_quack(): print("...")


def squeak(): print("squeak")


class NormalQuack:
    def __init__(self):
        self.cycle = ["quack", "quaaaak", "QUACK!"]
        self.idx = -1

    def do_quack(self):
        self.idx = (self.idx + 1) % len(self.cycle)
        print(self.cycle[self.idx])


class Duck:
    def __init__(self, quack_behavior=no_quack):
        self.quack_behavior = quack_behavior

    def quack(self):
        self.quack_behavior()


class RubberDuck(Duck):
    def __init__(self):
        super().__init__(squeak)


class AnconaDuck(Duck):
    def __init__(self):
        super().__init__(NormalQuack().do_quack)


if __name__ == '__main__':
    for duck in [AnconaDuck(), AnconaDuck(), RubberDuck()]:
        duck.quack()
        duck.quack()
        duck.quack()
        duck.quack()

File: src/strategy.rs
//! Strategy Pattern
//! ================
//!

trait QuackBehavior: 'static {
    fn do_quack(&mut self);
}

struct Duck {
    kind: &'static str,
    quack_behavior: Box<dyn QuackBehavior>,
}

impl Duck {
    pub fn new(kind: &'static str, quack_behavior: impl QuackBehavior) -> Self {
        Duck {
            kind,
            quack_behavior: Box::new(quack_behavior),
        }
    }

    pub fn quack(&mut self) {
        self.quack_behavior.do_quack()
    }
}

struct Quack;
impl QuackBehavior for Quack {
    fn do_quack(&mut self) {
        println!("quack")
    }
}

struct Squeak;
impl QuackBehavior for Squeak {
    fn do_quack(&mut self) {
        println!("squeak")
    }
}

#[test]
fn duck_concert() {
    let mut ducks = vec![
        Duck::new("Ancona Duck", Quack),
        Duck::new("Rubber Duck", Squeak),
    ];

    for duck in &mut ducks {
        duck.quack();
        duck.quack();
    }
}
